### ISSUE TITLE
Photometry discussion

### ISSUE THREAD

**Issue by rachel3834** [pyDANDIA/pyDANDIA#10]
I've implemented v1 of the PSF-fitting photometry code, based on the photutils package (which applies algorithms derived from DAOphot).  

This has raised a couple of questions:
- The photometry output does not appear to contain photometric uncertainties on the stellar flux measurements. 
- The PSF model used appears to be restricted to a Gaussian-based model.

My tests using the IntegratedGaussianPRF PSF model have not produced adequate photometry yet (see attached PSF-subtracted image for example).  

So: I've contacted the developer of photutils to ask about these points.  It may be that like much of astropy, its a great idea that's a work-in-progress; not ready to use yet.  

![residuals_image_test](https://user-images.githubusercontent.com/10391660/31799607-eb912a5c-b4ef-11e7-8ca7-10f6dadf57a1.png)

If so, then it may be necessary to build our own photometry package to use in the meantime.

**Comment by rachel3834** [comment]
For the record, this code is currently on the 'photometry' branch.

**Comment by mpgh** [comment]
Have you tried [BasicPSFPhotometry](http://photutils.readthedocs.io/en/stable/api/photutils.psf.BasicPSFPhotometry.html#photutils.psf.BasicPSFPhotometry)?

Nevertheless, you are right that it might be worth considerung a different method. It will also be used for the final diffimg photometry and could easily become a limiting factor for providing good photometry.

**Comment by rachel3834** [comment]
Thanks, yes,  I did explore the BasicPSF option, but the residuals were worse.  As you know, the iterative algorithm is more appropriate for crowded fields because it attempts to remove the companion stars before re-building the PSF.

**Comment by ytsapras** [comment]
We need to run some more tests and understand this behaviour. Should this
fail to meet our needs, we can consider using one of the existing
photometry packages or implementing our own.


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by ebachelet** [comment]
Thanks Rachel. That looks  bad to me. This white/black/white annealing looks like a bad PSF shape, as you suggested. The Moffat vs Gaussian looks definetely something important here. Moreover, the fact to not have uncertainties is something really anyoing me, since in my mind a measurement without uncertainties is meaningless. As you said, I also used BasicPSFPhotometry and was not really pleased by the results.
I used our PSF packages to run PSF fits examples I circulate some times ago. I have to say the use was pretty straightforward, the fitting process was easy (basicaly scipy.optimize.leastsq) and results convincing. I am sure your case is more difficult, but I am also pretty sure implementing our own stuff should not be to bad. I can help on this if needed.

**Comment by mpgh** [comment]
The original DANDIA version also had some issues with stage4. The resulting residual image also contains some affected areas and some objects that haven't been identified. Nevertheless, we still want to see an improvement. The most important change is that the old stage4 was relying on an empiric PSF model.

![image](https://user-images.githubusercontent.com/14907039/31898223-dede27d4-b818-11e7-85fe-0af63f182664.png)

**Comment by rachel3834** [comment]
Thanks everyone for your comments.  While I was away this weekend, I had the following responses from the package's developers:

Hi Rachel,

Thank you for your interest in photutils and providing feedback on the package.  It is very useful.  I'm including Zé Vinícius on this email.  He wrote most of the PSF fitting tools in photutils, including adding the PSF-fitting uncertainties (https://github.com/astropy/photutils/pull/516).

To answer your first question, the uncertainties should be returned in the photometry table in the column names that end in "_unc".  However, it requires using an astropy fitter than can return uncertainties, like LevMarLSQFitter.  I made a simple example here:

https://gist.github.com/larrybradley/e23d905cd69ca44d032541f69c2b5433

This feature should be better documented, so I created a new issue (https://github.com/astropy/photutils/issues/605).

There is not an explicit 2D Moffat PSF class, but I think you could create a Fittable2DModel (http://photutils.readthedocs.io/en/stable/api/photutils.psf.FittableImageModel.html#photutils.psf.FittableImageModel) that is a discretized and normalized astropy Moffat2D model (http://docs.astropy.org/en/stable/api/astropy.modeling.functional_models.Moffat2D.html).  I'm not sure if that is the best solution, but Zé can probably best answer this question.

Thanks again for your feedback!  It is much appreciated.  In the future, I have a gentle request that you please post these types of questions on the photutils issue tracker (https://github.com/astropy/photutils/issues).  That way if I'm not available, someone else may be able to respond.  Also, since it's publicly viewable and searchable, it can help others who are possibly having the same trouble and it prevents issues and/or bug reports from being reported multiple times.

Thanks again!
Larry Bradley
---------------------------

Hi Rachel,

Please refer to Larry's example at: https://gist.github.com/larrybradley/e23d905cd69ca44d032541f69c2b5433

There is not an explicit 2D Moffat PSF class, but I think you could create a Fittable2DModel (http://photutils.readthedocs.io/en/stable/api/photutils.psf.FittableImageModel.html#photutils.psf.FittableImageModel) that is a discretized and normalized astropy Moffat2D model (http://docs.astropy.org/en/stable/api/astropy.modeling.functional_models.Moffat2D.html).  I'm not sure if that is the best solution, but Zé can probably best answer this question.

I can't think of a better way!

Thank you,
Zé

**Comment by ebachelet** [comment]
Nice. But if we have to create a 2D moffat model in this module and use a lstsq fitter, we have already this implemented in the psf module. @ytsapras use this in his stage1, and it works well and return the fit covariance. From tests I made, that worked well.

**Comment by mpgh** [comment]
That sounds encouraging, so perhaps we should give Moffat a chance. Especially since the empirical PSF model has issues, too. Based on what I see, I would guess that we need a brightness dependent PSF model which looks more like a Moffat on some wings and more like a Lorentzian or Gaussian on others...

**Comment by rachel3834** [comment]
Yes, if we would need to re-implement the Moffat function anyway to use photutils, I agree it may be better to proceed with the code we already have.  

To answer @ytsapras's comment about implementing an alternative existing photometry code - I agree that trying not to "re-invent the wheel" is a good idea which is why I thought this avenue was worth investigating before we finalize the code.  

However, photutils appears to be the best-prepared existing photometry package - by which I mean that it is in a form which can be easily integrated into our Python pipeline, version controlled etc and implements photometry routines of Stetson, which are "industry standard", tried and tested.  All others that I know of are external packages (introducing separate version control issues and the necessity for external system calls) or not as well supported (introducing the possibility of being left with workarounds or bugs).  

I've followed up on the comments from photutils authors, since the example they sent didn't work for me.  This suggests that the lack of photometric uncertainties may be a local-install issue rather than a problem with the package.  But that doesn't change the issue with the PSF model.  

In parallel however, I will proceed using our own PSF fitting routines.

**Comment by rachel3834** [comment]
OK another update from the photutils team: apparently photometric uncertainties were only added in a recent version of photutils, which has not yet made it to an official release.

**Comment by danlegend5** [comment]
Could this be of use?

http://adsabs.harvard.edu/abs/2017RMxAC..49..146M

**Comment by rachel3834** [comment]
Worth checking out certainly, thanks for the link.
Rache

**Comment by ebachelet** [comment]
Fixed.
